web/packages/teleterm/src/ui/AccessRequestCheckout/useAccessRequestCheckout.ts

```diff
@@ -385,3 +375,13 @@ export default function useAccessRequestCheckout() {
     setRequestTTL,
   };
 }
+
+type PendingListItemWithOriginalItem = Omit<PendingListItem, 'kind'> & {
+  kind: ResourceKind;
+  originalItem:
+    | {
+        kind: 'role';
+        role: string;
+      }
+    | { kind: 'resource'; resource: ResourceRequest };
+};
```

Comment:
Something like this maybe?

```typescript
type PendingListItemWithOriginalItem = Omit<PendingListItem, 'kind'> &
  (
    | {
        kind: Exclude<ResourceKind, 'role'>;
        originalItem: ResourceRequest;
      }
    | {
        kind: 'role';
      }
  );
```

<details>
<summary>Patch</summary>

```diff
diff --git a/web/packages/teleterm/src/ui/AccessRequestCheckout/AccessRequestCheckout.tsx b/web/packages/teleterm/src/ui/AccessRequestCheckout/AccessRequestCheckout.tsx
index b676138087..5e2678e967 100644
--- a/web/packages/teleterm/src/ui/AccessRequestCheckout/AccessRequestCheckout.tsx
+++ b/web/packages/teleterm/src/ui/AccessRequestCheckout/AccessRequestCheckout.tsx
@@ -160,7 +160,7 @@ export function AccessRequestCheckout() {
                       case 'role':
                         break;
                       default:
-                        c.kind satisfies never;
+                        c satisfies never;
                     }
                     return resource;
                   })
diff --git a/web/packages/teleterm/src/ui/AccessRequestCheckout/useAccessRequestCheckout.ts b/web/packages/teleterm/src/ui/AccessRequestCheckout/useAccessRequestCheckout.ts
index 2050888705..a6ec0516c8 100644
--- a/web/packages/teleterm/src/ui/AccessRequestCheckout/useAccessRequestCheckout.ts
+++ b/web/packages/teleterm/src/ui/AccessRequestCheckout/useAccessRequestCheckout.ts
@@ -170,10 +170,6 @@ export default function useAccessRequestCheckout() {
             id: role,
             name: role,
             clusterName,
-            originalItem: {
-              kind: 'role',
-              role: role,
-            },
           });
         });
         break;
@@ -186,7 +182,7 @@ export default function useAccessRequestCheckout() {
             kind,
             id,
             name,
-            originalItem: { kind: 'resource', resource: resourceRequest },
+            originalItem: resourceRequest,
             clusterName: ctx.clustersService.findClusterByResource(
               resourceRequest.resource.uri
             )?.name,
@@ -204,15 +200,17 @@ export default function useAccessRequestCheckout() {
     return workspaceAccessRequest.getCollapsed();
   }
 
-  async function toggleResource({
-    originalItem,
-  }: PendingListItemWithOriginalItem) {
-    if (originalItem.kind === 'role') {
-      await workspaceAccessRequest.addOrRemoveRole(originalItem.role);
+  async function toggleResource(
+    pendingListItem: PendingListItemWithOriginalItem
+  ) {
+    if (pendingListItem.kind === 'role') {
+      await workspaceAccessRequest.addOrRemoveRole(pendingListItem.id);
       return;
     }
 
-    await workspaceAccessRequest.addOrRemoveResource(originalItem.resource);
+    await workspaceAccessRequest.addOrRemoveResource(
+      pendingListItem.originalItem
+    );
   }
 
   function getAssumedRequests() {
@@ -385,12 +383,13 @@ export default function useAccessRequestCheckout() {
   };
 }
 
-type PendingListItemWithOriginalItem = Omit<PendingListItem, 'kind'> & {
-  kind: ResourceKind;
-  originalItem:
+type PendingListItemWithOriginalItem = Omit<PendingListItem, 'kind'> &
+  (
+    | {
+        kind: Exclude<ResourceKind, 'role'>;
+        originalItem: ResourceRequest;
+      }
     | {
         kind: 'role';
-        role: string;
       }
-    | { kind: 'resource'; resource: ResourceRequest };
-};
+  );

```
</details>

It's not ideal, because you can still write code like this:

```typescript
const node: PendingListItemWithOriginalItem = {
  kind: 'node',
  id: 'foo',
  name: 'bar',
  clusterName: 'hello',
  originalItem: { kind: 'database', resource: { uri: '' } }
}
```

But at least it rules out code like this:

```typescript
const item: PendingListItemWithOriginalItem = {
  kind: 'node',
  id: 'foo',
  name: 'bar',
  clusterName: 'hello',
  originalItem: { kind: 'role', role: 'baz' },
};
```

Reply:
That's better, thanks.